File: two_sum.py
class Solution:
    def twoSum(self,num,target):
        dic={}
        for index in range(len(num)):
            if num[index] not in dic:
                
                dic[num[index]]=[index]
            else:
                dic[num[index]].append(index)           
            
                
        print dic
        start=0
        num.sort()
        end=len(num)-1
        while start<end:
            if num[start]+num[end]>target:
                end-=1
            elif num[start]+num[end]<target:
                start+=1
            else:
                if  num[start]==num[end]:
                    return(dic[num[start]][0]+1,dic[num[end]][1]+1)
                else:
                    return (dic[num[start]][0]+1,dic[num[end]][0]+1)
        return
    def twoSum_1(self,num,target):
        dic={}
        length=len(num)
        for index in range(length):
            if num[index] not in dic:
                dic[num[index]]=index+1
        for i in range(length):
            if dic.get(target-num[i]):
                if i+1!=dic[target-num[i]]:
                    return  i+1,dic[target-num[i]]         


if __name__=='__main__':
    sol=Solution()
    import random
    num=[3,2,4]
    print sol.twoSum_1(num,6)

File: Algorithm/CPP/two_sum.cpp
# include <iostream>
# include <string>
# include <stdlib.h>
# include <algorithm>
# include <vector>
# include <unordered_map>
# include <map>
using namespace std;

class Solution {
public:
	vector<int> twoSum1(vector<int>& nums, int target) 
	{
	unordered_map<int, int> hash;
	vector<int> res;
	for (auto i = 0; i < nums.size(); ++i)
	{
		if (hash.find(nums[i]) == hash.end())
			hash[nums[i]] = i;
	}

	for (int i = nums.size() - 1; i > -1; --i)
	{
		if (hash.find(target - nums[i]) != hash.end())
		{
			if ((i) != (hash[target - nums[i]]))
			{
				res.push_back(hash[target - nums[i]] + 1);
				res.push_back(i + 1);
				return res;
			}
		}
	}
	}

	vector<int> twoSum ( vector<int>& nums, int target )
	{
		unordered_map<int, int> dic;
		vector <int> res;
		for (auto i = 0; i < nums.size(); ++i)
		{
			if (dic.find(target - nums[i]) != dic.end())
			{
				res.push_back(dic[target - nums[i]] + 1);
				res.push_back(i + 1);
				return res;
			}
			dic[nums[i]] = i;
		}
	}
};


int main()
{

	Solution sol;
	vector<int> numbers = { 2, 7, 11, 15 };
	auto target = 9;
	for (auto s : sol.twoSum(numbers, target))
		cout << s << " ";
	cout << endl;
	cin.get();
	return 0;
}
